### PR TITLE
Add meta to to_dask_array

### PR DESCRIPTION
**Pull request by kylejn27**
**What**

Adds optional metadata param to `to_dask_array()`

**Why**

Resolving this issue:

https://github.com/dask/dask/issues/6349

- [X] Tests added / passed
- [X] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by quasiben** [comment]
@pentschev in case you have thoughts about `meta` here

**Review by pentschev — COMMENTED** [review]
Changes look fine to me. I added only a small suggestion on a test, but it's good to go from my side.

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -2659,6 +2660,10 @@ def test_to_dask_array(as_frame, lengths):
 
     assert result.chunks == expected_chunks
 
+    if meta is not None:
+        assert result.dtype == meta.dtype
+        assert result.dtype != "i4"
```
[the remark this answers is not in the document]
Comment (kylejn27) [reply dask/dataframe/tests/test_dataframe.py:2665]: So, line 2665 was added in mistake. I added it during development as a quick test and meant to remove it. Are you suggesting that lines 2663 through 2665 be removed entirely? I suppose I could, leaving 2653 just testing that meta parameter exists, but the purpose of the test was to ensure that the metadata passed into `to_dask_array` was changing the datatype of the resulting array. Maybe a better test would be to compare `result._meta` to `meta` and check for equality

[the remark this answers is not in the document]
Comment (pentschev) [reply dask/dataframe/tests/test_dataframe.py:2665]: I meant only removing line 2665 and keep the rest, sorry for the confusion. Checking meta for equality could be useful too.

**Comment by jsignell** [comment]
Just a note that you'll probably want to add meta to the methods in dask/arrray/core.py as well.

**Review by pentschev — APPROVED** [review]
LGTM, thanks for the work on this @kylejn27 !

**Comment by mrocklin** [comment]
Sorry to come in late here, but I don't understand why we would need `meta=` keywords for reductions like `any` and `all` or `min` and `max`.  I can't think of a situation where a user would need to provide these and we wouldn't be able to provide these for them.

Looking through the previous issue I find that there is a bug with 

```python
a = da.ones((4,), dtype='datetime64[ns]')
a.min()
```

The solution presented here feels a bit like we're creating space for a workaround for some advanced users, perhaps at the cost of complexity, while leaving tha majority of users behind.

There are a few things that we could do instead:

1.  The min/... functions could determine the meta (presumably it's exactly the same as the input array but with a different shape) and send that instead.  This removes the burden from the user.
2.  Even better, we could figure out why we're calling `sum` when calling `min` and see why that is causing issues lower down so that not every function  needed to determine its meta explicitly.  Has anyone gone through the traceback yet to determine what `reduction` is doing here and why it's calling `sum`?  I wonder if there is an alternative there.

**Comment by mrocklin** [comment]
(to be clear, I don't have an issue with adding meta to `to_dask_array`.  It's the .min/max/sum methods that concern me here.)

**Comment by kylejn27** [comment]
@mrocklin I think this conversation explains a little bit about what's going on https://github.com/dask/dask/issues/6354#issuecomment-696968527

I'm totally cool with removing meta from .min/max/sum methods, that was a misinterpretation on my part but it seemed that it was a wanted contribution so I kept it in anyways

The real issue is that sum is being called on the reduced metadata to reduce the dimensionality of the metadata. Without it, the metadata has a different dimensionality than the result

**Comment by mrocklin** [comment]
I wonder if we can replace the use of sum with an understanding of the axis
and keepdims arguments to achieve the same result.  This would probably
require a little bit of work on our part, but might reduce complexity for
others in the future.

**Comment by kylejn27** [comment]
> I wonder if we can replace the use of sum with an understanding of the axis
and keepdims arguments to achieve the same result.

Sounds like a great place to start. 

On the topic of meta in min/max/sum, I'll pull that out of this PR since it seems that it's not desired

**Comment by mrocklin** [comment]
Woot.  Thanks Kyle

File: dask/dataframe/tests/test_dataframe.py [hunk 2]
```diff
@@ -2640,16 +2640,17 @@ def test_to_dask_array_unknown(as_frame):
...
 
 
-@pytest.mark.parametrize("lengths", [[2, 3], True])
-@pytest.mark.parametrize("as_frame", [False, False])
-def test_to_dask_array(as_frame, lengths):
-    s = pd.Series([1, 2, 3, 4, 5], name="foo")
+@pytest.mark.parametrize("lengths", [[2, 3], True, True])
+@pytest.mark.parametrize("as_frame", [False, False, False])
+@pytest.mark.parametrize("meta", [None, None, np.array([], dtype="f4")])
+def test_to_dask_array(meta, as_frame, lengths):
```
Comment (mrocklin) [inline dask/dataframe/tests/test_dataframe.py:2646]: Given that these all have length three I suspect that your intention here is to run three instances of this test.  However, multiple parametrize calls produce cartesian products, so this is actually creating `3**3 == 27` instances of the same test.  Maybe you meant to use 

```python
@pytest.mark.parametrize("lengths,as_frame,meta", [(..., ..., ...), (..., ..., ...), ...])
```

Comment (kylejn27) [reply dask/dataframe/tests/test_dataframe.py:2646]: good catch. I also notice that as_frame is never True, which seems like it wasn't the original intention of the test. I can fix that as well

**Comment by mrocklin** [comment]
Thanks @kylejn27 .  This is in.

**Pull request merged**